compiler/src/iree/compiler/Dialect/Flow/Transforms/RegionOpUtils.cpp

```diff
@@ -528,6 +528,149 @@ wrapOpInDispatchRegion(RewriterBase &rewriter, Operation *op) {
+        Operation *producer = operand.getDefiningOp();
+        return producer && producer->getParentOfType<DispatchRegionOp>();
+      })) {
+    rewriter.setInsertionPoint(dispatchRegionOp);
+  } else if (llvm::all_of(op->getUsers(), [&](Operation *user) {
+               return isa<IREE::Flow::ReturnOp>(user);
+             })) {
+    rewriter.setInsertionPointAfter(dispatchRegionOp);
+  } else {
+    return rewriter.notifyMatchFailure(
+        op, "op has both operands and users insided of its dispatch");
+  }
+  Operation *hoistedOp = rewriter.clone(*op);
+  rewriter.setInsertionPoint(dispatchRegionOp);
+
+  // Get the new dispatch region result types without the hoisted op results.
+  auto dispatchReturnOp = cast<IREE::Flow::ReturnOp>(
+      dispatchRegionOp.getBody().front().getTerminator());
+  SmallVector<Type> newResultTypes;
+  SmallVector<OpOperand *> newDispatchReturnOperands;
+  for (OpOperand &operand : dispatchReturnOp->getOpOperands()) {
+    if (operand.get().getDefiningOp() != op) {
+      newResultTypes.push_back(operand.get().getType());
+      newDispatchReturnOperands.push_back(&operand);
+    }
+  }
+
+  // Replace op uses inside and outside of the dispatch region with the hoisted
+  // results.
+  bool yieldsResults = false;
+  for (OpResult result : op->getResults()) {
+    Value hoistedResult = hoistedOp->getResult(result.getResultNumber());
+    // Replace all results yielded by the dispatch region with the hoisted
+    // op results.
+    for (OpOperand &use : result.getUses()) {
+      if (!isa<IREE::Flow::ReturnOp>(use.getOwner())) {
+        continue;
+      }
+      yieldsResults = true;
+      Value dispResult = dispatchRegionOp.getResults()[use.getOperandNumber()];
+      rewriter.replaceAllUsesWith(dispResult, hoistedResult);
+    }
+    // Replace uses inside the dispatch region.
+    rewriter.replaceAllUsesWith(result, hoistedResult);
+  }
+  // If no results were yielded from `op`, then nothing more to do.
+  if (!yieldsResults) {
+    return hoistedOp;
+  }
+
+  // Add the operands of the `op` to the new return values of the dispatch.
+  SmallVector<Value> newReturnedValues = llvm::map_to_vector(
+      newDispatchReturnOperands, [](OpOperand *val) { return val->get(); });
+  SetVector<Value> newReturnedValueSet(newReturnedValues.begin(),
+                                       newReturnedValues.end());
+  SmallVector<OpOperand *> yieldedOperands;
+  SmallVector<SmallVector<Value>> newResultsDynamicDims;
+  for (OpOperand &operand : op->getOpOperands()) {
+    // Only need to yield operands defined in the dispatch region.
+    if (operand.get().getParentRegion() != &dispatchRegionOp.getBody()) {
+      continue;
+    }
+    // Skip already yielded operands.
+    if (newReturnedValueSet.contains(operand.get())) {
+      continue;
+    }
+    // Save operand and dynamic dims to add to the dispatch region.
+    yieldedOperands.push_back(&operand);
+    OpBuilder::InsertionGuard guard(rewriter);
```

Comment:
I'd expect a single `InsertionGuard` in a function because function usually should not be large. (Though, the length of the function is ~140 lines, which is large to me.) I find that tracking the insertion point in a function is very painful. It is adding complexity to my brain because it triggers me to think where the insertion point was. If you can refactor it into a couple functions, having insertion guard is reasonable to me. Otherwise, I'd suggest to remove all of them (except the first one) from the function. In the scope of the function, I don't have concerns about the state of insertion point. What do you think?